README.md
# SlidingLayout

SlidingLayout是一种Android平台的View控件，可以帮助你实现类似微信网页浏览的下拉功能，也可以帮助你实现类似iOS中UITableView的下拉上拉弹跳的果冻效果。

SlidingLayout完美兼容Android自带库和兼容库的所有View组件，包括RecyclerView、ListView、ScrollView以及WebView等等。

SlidingLayout简单易用，最低支持Android api v9。

Project site： <https://github.com/HomHomLin/SlidingLayout>.

Demo: <https://github.com/HomHomLin/SlidingLayout/tree/master/demo>.

最新版本:v0.9.0

## 效果图：

![p1](https://raw.githubusercontent.com/HomHomLin/SlidingLayout/master/pic/demo.gif)

![p2](https://raw.githubusercontent.com/HomHomLin/SlidingLayout/master/pic/list.gif)

![p3](https://raw.githubusercontent.com/HomHomLin/SlidingLayout/master/pic/webview.gif)

## 导入项目

**Gradle dependency:**
``` groovy
compile 'homhomlin.lib:sldinglayout:0.9.0'
```

or

**Maven dependency:**
``` xml
<dependency>
  <groupId>homhomlin.lib</groupId>
  <artifactId>sldinglayout</artifactId>
  <version>0.9.0</version>
</dependency>
```

**依赖:**

如果你的项目需要支持API V9，你需要添加以下依赖：

``` groovy
compile 'com.nineoldandroids:library:2.4.0'
```

## 用法

SlidingLayout的使用非常简单，你只需要将你想实现的控件在XML布局中嵌套进SlidingLayout即可，如你需要让ListView实现果冻效果：

### 1.创建背景View的xml

``` xml
<?xml version="1.0" encoding="utf-8"?>
<TextView xmlns:android="http://schemas.android.com/apk/res/android"
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    android:background="#8c8c8e"
    android:gravity="top"
    android:textSize="12sp"
    android:textColor="#f5f3f3"
    android:padding="16dp"
    android:text="developed by HomhomLin"/>
```

### 2.将你的控件放进SlidingLayout中

注意布局需要res-auto命名空间，注意将自己的控件设置一个背景，否则会将背景View透视出来。

```xml
<?xml version="1.0" encoding="utf-8"?>
<lib.homhomlib.design.SlidingLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:id="@+id/slidingLayout"
    android:layout_width="match_parent"
    android:layout_height="match_parent"
    app:sliding_mode="both"
    app:background_view="@layout/view_bg">
    <!--background_view为你的背景布局-->
    <ListView
        android:id="@+id/listview"
        android:background="#ffffff"
        android:layout_width="match_parent"
        android:layout_height="match_parent">
    </ListView>
</lib.homhomlib.design.SlidingLayout>
```

### 3.没有第三步啦！

运行即可看到效果！具体其他的内容可以看Demo。

## XML样式参数

 * `background_view` 背景view
 * `sliding_mode` 滑动模式，both为上下可弹跳，top为顶部弹跳，bottom为底部弹跳，默认为both
 * `sliding_pointer_mode` 手指模式，one为只识别一个手指，more为支持多指滑动，默认为more
 * `top_max` 当滑动模式为top时才有效，用于可滑动的最大距离，如"top_max:200dp"，默认为-1（不限制）

## 常用API

 * `public void setSlidingOffset(float slidingOffset)` 设置控件的滑动阻力，有效值为0.1F~1.0F，值越小阻力越大，默认为0.5F
 * `public void setTargetView(View view)` 设置控件的前景View
 * `public void setBackgroundView(View view)` 设置控件的背景View
 * `public void setSlidingListener(SlidingListener slidingListener)` 给控件设置监听，可以监听滑动情况
 * `public void setSlidingMode(int mode)` 设置滑动模式
 * `public void setSlidingDistance(int max)` 设置最大滑动距离，仅在top模式下有效

## Developed By

 * Linhonghong - <linhh90@163.com>

## License
Copyright 2016 LinHongHong

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

ChangeLog.md
0.8.0

update:

①增加多指识别，现在多个手指可以累积滑动效果，默认开启
②滑动回弹时间缩短100ms

0.8.1

update:
①fix bugs
②增加最大下拉距离,top_max
③增加代理onTouchListener

0.9.0

update:
滑动阻力效果替换为越拉越吃力

app/src/test/java/lib/homhomlib/design/ExampleUnitTest.java
package lib.homhomlib.design;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * To work on unit tests, switch the Test Artifact in the Build Variants view.
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

library/src/main/java/lib/homhomlib/design/Instrument.java
package lib.homhomlib.design;

import android.view.View;

import com.nineoldandroids.view.ViewHelper;

/**
 * Created by Linhh on 16/5/24.
 */
public class Instrument {
    private static Instrument mInstrument;
    public static Instrument getInstance(){
        if(mInstrument == null){
            mInstrument = new Instrument();
        }
        return mInstrument;
    }

    public float getTranslationY(View view){
        if(view == null){
            return 0;
        }
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.HONEYCOMB) {
            return view.getTranslationY();
        }else{
            return ViewHelper.getTranslationY(view);
        }
    }

    public void slidingByDelta(View view ,float delta){
        if(view == null){
            return;
        }
        view.clearAnimation();
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.HONEYCOMB) {
            view.setTranslationY(delta);
        }else{
            ViewHelper.setTranslationY(view, delta);
        }
    }

    public void slidingToY(View view ,float y){
        if(view == null){
            return;
        }
        view.clearAnimation();
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.HONEYCOMB) {
            view.setY(y);
        }else{
            ViewHelper.setY(view, y);
        }
    }

    public void reset(View view, long duration){
        if(view == null){
            return;
        }
        view.clearAnimation();
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.HONEYCOMB) {
            android.animation.ObjectAnimator.ofFloat(view, "translationY", 0F).setDuration(duration).start();
        }else{
            com.nineoldandroids.animation.ObjectAnimator.ofFloat(view, "translationY", 0F).setDuration(duration).start();
        }
    }

    public void smoothTo(View view ,float y, long duration){
        if(view == null){
            return;
        }
        view.clearAnimation();
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.HONEYCOMB) {
            android.animation.ObjectAnimator.ofFloat(view, "translationY", y).setDuration(duration).start();
        }else{
            com.nineoldandroids.animation.ObjectAnimator.ofFloat(view, "translationY", y).setDuration(duration).start();
        }
    }
}


app/src/main/java/lib/homhomlib/design/ui/MainActivity.java
package lib.homhomlib.design.ui;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.webkit.WebView;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Button webView = (Button) this.findViewById(R.id.btnWebView);
        webView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(MainActivity.this, WebViewActivity.class));
            }
        });
        Button listView = (Button)this.findViewById(R.id.btnListView);
        listView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(MainActivity.this,ListViewActivity.class));
            }
        });
        Button recyclerView = (Button)this.findViewById(R.id.btnRecyclerView);
        recyclerView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(MainActivity.this,RecyclerViewActivity.class));
            }
        });
    }
}


library/src/main/java/lib/homhomlib/design/SlidingLayout.java
package lib.homhomlib.design;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.support.v4.view.MotionEventCompat;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.widget.AbsListView;
import android.widget.FrameLayout;

/**
 * github:https://github.com/HomHomLin/SlidingLayout
 * Created by Linhh on 16/4/12.
 */
public class SlidingLayout extends FrameLayout{

    private int mTouchSlop;//系统允许最小的滑动判断值
    private int mBackgroundViewLayoutId = 0;

    private View mBackgroundView;//背景View
    private View mTargetView;//正面View

    private boolean mIsBeingDragged;
    private float mInitialDownY;
    private float mInitialMotionY;
    private float mLastMotionY;
    private int mActivePointerId = INVALID_POINTER;

    private float mSlidingOffset = 0.5F;//滑动阻力系数

    private static final int RESET_DURATION = 200;
    private static final int SMOOTH_DURATION = 1000;

    public static final int SLIDING_MODE_BOTH = 0;
    public static final int SLIDING_MODE_TOP = 1;
    public static final int SLIDING_MODE_BOTTOM = 2;

    public static final int SLIDING_POINTER_MODE_ONE = 0;
    public static final int SLIDING_POINTER_MODE_MORE = 1;

    private int mSlidingMode = SLIDING_MODE_BOTH;

    private int mSlidingPointerMode = SLIDING_POINTER_MODE_MORE;

    private static final int INVALID_POINTER = -1;

    private SlidingListener mSlidingListener;

    public static final int STATE_SLIDING = 2;
    public static final int STATE_IDLE = 1;

    private int mSlidingTopMaxDistance = SLIDING_DISTANCE_UNDEFINED;

    public static final int SLIDING_DISTANCE_UNDEFINED = -1;

    private OnTouchListener mDelegateTouchListener;

    public interface SlidingListener{
        //不能操作繁重的任务在这里
        public void onSlidingOffset(View view, float delta);
        public void onSlidingStateChange(View view ,int state);
        public void onSlidingChangePointer(View view, int pointerId);
    }

    public SlidingLayout(Context context) {
        this(context, null);
    }

    public SlidingLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SlidingLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs){
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.SlidingLayout);
        mBackgroundViewLayoutId = a.getResourceId(R.styleable.SlidingLayout_background_view, mBackgroundViewLayoutId);
        mSlidingMode = a.getInteger(R.styleable.SlidingLayout_sliding_mode,SLIDING_MODE_BOTH);
        mSlidingPointerMode = a.getInteger(R.styleable.SlidingLayout_sliding_pointer_mode,SLIDING_POINTER_MODE_MORE);
        mSlidingTopMaxDistance = a.getDimensionPixelSize(R.styleable.SlidingLayout_top_max,SLIDING_DISTANCE_UNDEFINED);
        a.recycle();
        if(mBackgroundViewLayoutId != 0){
            View view = View.inflate(getContext(), mBackgroundViewLayoutId, null);
            setBackgroundView(view);
        }
        mTouchSlop = ViewConfiguration.get(getContext()).getScaledTouchSlop();
    }

    public void setBackgroundView(View view){
        if(mBackgroundView != null){
            this.removeView(mBackgroundView);
        }
        mBackgroundView = view;
        this.addView(view, 0);
    }

    public View getBackgroundView(){
        return this.mBackgroundView;
    }

    public void setSlidingDistance(int distance){
        this.mSlidingTopMaxDistance = distance;
    }

    public int setSlidingDistance(){
        return this.mSlidingTopMaxDistance;
    }

    /**
     * 获得滑动幅度
     * @return
     */
    public float getSlidingOffset(){
        return this.mSlidingOffset;
    }

    /**
     * 设置滑动幅度
     * @param slidingOffset
     */
    public void setSlidingOffset(float slidingOffset){
        this.mSlidingOffset = slidingOffset;
    }

    public void setSlidingListener(SlidingListener slidingListener){
        this.mSlidingListener = slidingListener;
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        //实际上整个layout只能存在一个背景和一个前景才有用途
//        if(getChildCount() > 2){
//
//        }
        if (getChildCount() == 0) {
            return;
        }
        if (mTargetView == null) {
            ensureTarget();
        }
        if (mTargetView == null) {
            return;
        }
    }

    private void ensureTarget() {
        if (mTargetView == null) {
            mTargetView = getChildAt(getChildCount() - 1);
        }
    }

    public void setTargetView(View view){
        if(mTargetView != null){
            this.removeView(mTargetView);
        }
        mTargetView = view;
        this.addView(view);
    }

    @Override
    public void setOnTouchListener(OnTouchListener l) {
//        super.setOnTouchListener(l);
        mDelegateTouchListener = l;
    }

    public View getTargetView(){
        return this.mTargetView;
    }

    public float getSlidingDistance(){
        return getInstrument().getTranslationY(getTargetView());
    }

    public Instrument getInstrument(){
        return Instrument.getInstance();
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {

        ensureTarget();

        final int action = MotionEventCompat.getActionMasked(ev);

        //判断拦截
        switch (action) {
            case MotionEvent.ACTION_DOWN:
//                Log.i("onInterceptTouchEvent", "down");
                mActivePointerId = MotionEventCompat.getPointerId(ev, 0);
                mIsBeingDragged = false;
                final float initialDownY = getMotionEventY(ev, mActivePointerId);
                if (initialDownY == -1) {
                    return false;
                }
                mInitialDownY = initialDownY;
                break;

            case MotionEvent.ACTION_MOVE:
                if (mActivePointerId == INVALID_POINTER) {
//                    Log.e(LOG_TAG, "Got ACTION_MOVE event but don't have an active pointer id.");
                    return false;
                }

                final float y = getMotionEventY(ev, mActivePointerId);
                if (y == -1) {
                    return false;
                }

                if(y > mInitialDownY) {
                    //判断是否是上拉操作
                    final float yDiff = y - mInitialDownY;
                    if (yDiff > mTouchSlop && !mIsBeingDragged && !canChildScrollUp()) {
                        mInitialMotionY = mInitialDownY + mTouchSlop;
                        mLastMotionY = mInitialMotionY;
                        mIsBeingDragged = true;
                    }
                }else if(y < mInitialDownY){
                    //判断是否是下拉操作
                    final float yDiff = mInitialDownY - y;
                    if (yDiff > mTouchSlop && !mIsBeingDragged && !canChildScrollDown()) {
                        mInitialMotionY = mInitialDownY + mTouchSlop;
                        mLastMotionY = mInitialMotionY;
                        mIsBeingDragged = true;
                    }
                }
                break;

            case MotionEventCompat.ACTION_POINTER_UP:
                break;

            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
//                Log.i("onInterceptTouchEvent", "up");
                mIsBeingDragged = false;
                mActivePointerId = INVALID_POINTER;
                break;
        }

        return mIsBeingDragged;
    }

    private float getMotionEventY(MotionEvent ev, int activePointerId) {
        final int index = MotionEventCompat.findPointerIndex(ev, activePointerId);
        if (index < 0) {
            return -1;
        }
        return MotionEventCompat.getY(ev, index);
    }

    /**
     * 判断View是否可以上拉
     * @return canChildScrollUp
     */
    public boolean canChildScrollUp() {
        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            if (mTargetView instanceof AbsListView) {
                final AbsListView absListView = (AbsListView) mTargetView;
                return absListView.getChildCount() > 0
                        && (absListView.getFirstVisiblePosition() > 0 || absListView.getChildAt(0)
                        .getTop() < absListView.getPaddingTop());
            } else {
                return ViewCompat.canScrollVertically(mTargetView, -1) || mTargetView.getScrollY() > 0;
            }
        } else {
            return ViewCompat.canScrollVertically(mTargetView, -1);
        }
    }

    /**
     * 判断View是否可以下拉
     * @return canChildScrollDown
     */
    public boolean canChildScrollDown() {
        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            if (mTargetView instanceof AbsListView) {
                final AbsListView absListView = (AbsListView) mTargetView;
                return absListView.getChildCount() > 0 && absListView.getAdapter() != null
                        && (absListView.getLastVisiblePosition() < absListView.getAdapter().getCount() - 1 || absListView.getChildAt(absListView.getChildCount() - 1)
                        .getBottom() < absListView.getPaddingBottom());
            } else {
                return ViewCompat.canScrollVertically(mTargetView, 1) || mTargetView.getScrollY() > 0;
            }
        } else {
            return ViewCompat.canScrollVertically(mTargetView, 1);
        }
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {

        return super.dispatchTouchEvent(event);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if(mDelegateTouchListener != null && mDelegateTouchListener.onTouch(this,event)){
            return true;
        }
        switch (event.getAction()){
            case MotionEvent.ACTION_DOWN:
//                Log.i("onTouchEvent", "down");
                break;
            case MotionEvent.ACTION_MOVE:
                float delta = 0.0f;
                float movemment = 0.0f;
                if(mSlidingPointerMode == SLIDING_POINTER_MODE_MORE) {
                    //homhom:it's different betweenn more than one pointer
                    int activePointerId = MotionEventCompat.getPointerId(event, event.getPointerCount() - 1);
                    if (mActivePointerId != activePointerId) {
                        //change pointer
//                    Log.i("onTouchEvent","change point");
                        mActivePointerId = activePointerId;
                        mInitialDownY = getMotionEventY(event, mActivePointerId);
                        mInitialMotionY = mInitialDownY + mTouchSlop;
                        mLastMotionY = mInitialMotionY;
                        if (mSlidingListener != null) {
                            mSlidingListener.onSlidingChangePointer(mTargetView, activePointerId);
                        }
                    }

                    //pointer delta
//                    delta = getInstrument().getTranslationY(mTargetView)
//                            + ((getMotionEventY(event, mActivePointerId) - mLastMotionY))
//                            / mSlidingOffset;

                    delta = getMotionEventY(event, mActivePointerId) - mLastMotionY;

                    //滑动阻力计算
//                    float tempOffset = getInstrument().getTranslationY(mTargetView)
//                            + delta;

                    float tempOffset = 1 - (Math.abs(getInstrument().getTranslationY(mTargetView)
                            + delta) / mTargetView.getMeasuredHeight());

                    delta = getInstrument().getTranslationY(mTargetView)
                            + delta * mSlidingOffset * tempOffset;

                    mLastMotionY = getMotionEventY(event, mActivePointerId);

                    //used for judge which side move to
                    movemment = getMotionEventY(event, mActivePointerId) - mInitialMotionY;
                }else {
                    float tempOffset = 1 - Math.abs(getInstrument().getTranslationY(mTargetView) / mTargetView.getMeasuredHeight());

                    delta = (event.getY() - mInitialMotionY) * mSlidingOffset * tempOffset;
                    //used for judge which side move to
                    movemment = event.getY() - mInitialMotionY;
                }

                float distance = getSlidingDistance();

                switch (mSlidingMode){
                    case SLIDING_MODE_BOTH:
                        getInstrument().slidingByDelta(mTargetView, delta);
                        break;
                    case SLIDING_MODE_TOP:
                        if(movemment >= 0 || distance > 0){
                            //向下滑动
                            if(delta < 0 ){
                                //如果还往上滑，就让它归零
                                delta = 0;
                            }

                            if(mSlidingTopMaxDistance == SLIDING_DISTANCE_UNDEFINED || delta < mSlidingTopMaxDistance){
                                //滑动范围内 for todo
                            }else{
                                //超过滑动范围
                                delta = mSlidingTopMaxDistance;
                            }

                            getInstrument().slidingByDelta(mTargetView, delta);
                        }
                        break;
                    case SLIDING_MODE_BOTTOM:
                        if(movemment <= 0 || distance < 0){
                            //向上滑动
                            if(delta > 0 ){
                                //如果还往下滑，就让它归零
                                delta = 0;
                            }
                            getInstrument().slidingByDelta(mTargetView, delta);
                        }
                        break;
                }


                if(mSlidingListener != null){
                    mSlidingListener.onSlidingStateChange(this, STATE_SLIDING);
                    mSlidingListener.onSlidingOffset(this,delta);
                }

                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
//                Log.i("onTouchEvent", "up");
                if(mSlidingListener != null){
                    mSlidingListener.onSlidingStateChange(this, STATE_IDLE);
                }
                getInstrument().reset(mTargetView,RESET_DURATION);
                break;
        }
        //消费触摸
        return true;
    }

    public void setSlidingMode(int mode){
        mSlidingMode = mode;
    }

    public int getSlidingMode(){
        return mSlidingMode;
    }

    public void smoothScrollTo(float y){
        getInstrument().smoothTo(mTargetView, y, SMOOTH_DURATION);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
    }


    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if(mTargetView != null){
            mTargetView.clearAnimation();
        }
        mSlidingMode = 0;
        mTargetView = null;
        mBackgroundView = null;
        mSlidingListener = null;
    }
}


app/src/main/java/lib/homhomlib/design/ui/RecyclerViewActivity.java
package lib.homhomlib.design.ui;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

/**
 * Created by Linhh on 16/4/18.
 */
public class RecyclerViewActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_recyclerview);
//        View front = View.inflate(this,R.layout.view_front,null);
        RecyclerView listView = (RecyclerView) this.findViewById(R.id.recycler_view);
        listView.setLayoutManager(new LinearLayoutManager(this));
        listView.setHasFixedSize(true);
        listView.setAdapter(new Adapter());
//        mSlidingLayout.setFrontView(front);
    }

    class Adapter extends RecyclerView.Adapter<Adapter.ViewHolder>{


        class ViewHolder extends RecyclerView.ViewHolder{

            public ViewHolder(View itemView) {
                super(itemView);
            }
        }

        @Override
        public Adapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View convertView = View.inflate(RecyclerViewActivity.this, R.layout.list_item, null);

            return new ViewHolder(convertView);
        }

        @Override
        public void onBindViewHolder(Adapter.ViewHolder holder, int position) {

        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public int getItemCount() {
            return 60;
        }
    }
}


app/src/main/java/lib/homhomlib/design/ui/ListViewActivity.java
package lib.homhomlib.design.ui;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;

import lib.homhomlib.design.SlidingLayout;

/**
 * Created by Linhh on 16/4/15.
 */
public class ListViewActivity extends AppCompatActivity {
    private SlidingLayout mSlidingLayout;
    private final static String TAG = "ListViewActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_listview);
        mSlidingLayout = (SlidingLayout) findViewById(R.id.slidingLayout);
//        View front = View.inflate(this,R.layout.view_front,null);
        ListView listView = (ListView) this.findViewById(R.id.listview);
        listView.setAdapter(new Adapter());
        mSlidingLayout.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()){
                    case MotionEvent.ACTION_CANCEL:
                    case MotionEvent.ACTION_UP:
                        Log.i("onTouch","up");
                        return false;
                    default:
                        return false;
                }
            }
        });
        mSlidingLayout.setSlidingListener(new SlidingLayout.SlidingListener() {
            @Override
            public void onSlidingOffset(View view, float delta) {
                Log.i(TAG,"onSlidingOffset:" + delta);
            }

            @Override
            public void onSlidingStateChange(View view, int state) {
                Log.i(TAG,"onSlidingStateChange:" + state);
            }

            @Override
            public void onSlidingChangePointer(View view, int pointerId) {

            }
        });
//        mSlidingLayout.setFrontView(front);
    }

    class Adapter extends BaseAdapter{

        @Override
        public int getCount() {
            return 60;
        }

        @Override
        public Object getItem(int position) {
            return null;
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            if(convertView == null){
                convertView = View.inflate(ListViewActivity.this,R.layout.list_item,null);

            }
            return convertView;
        }
    }
}